src/avram/delete_operation.cr

```diff
@@ -0,0 +1,234 @@
+require "./validations"
+require "./callbacks"
+require "./define_attribute"
+require "./operation_errors"
+require "./param_key_override"
+require "./inherit_column_attributes"
+require "./needy_initializer_and_delete_methods"
+
+abstract class Avram::DeleteOperation(T)
+  include Avram::NeedyInitializerAndDeleteMethods
+  include Avram::DefineAttribute
+  include Avram::Validations
+  include Avram::OperationErrors
+  include Avram::ParamKeyOverride
+  include Avram::Callbacks
+  include Avram::InheritColumnAttributes
+
+  enum DeleteStatus
+    Deleted
+    DeleteFailed
+    Unperformed
+  end
+
+  macro inherited
+    @@permitted_param_keys = [] of String
+
+    @record : T
+    @params : Avram::Paramable
+    getter :record, :params
+    property delete_status : DeleteStatus = DeleteStatus::Unperformed
+  end
+
+  def self.delete(*args, **named_args, &block)
```

Comment:
Maybe this can be moved in to that `NeedyInitializerAndDeleteMethods` so it can generate both `delete` and `delete!` with or without the block?

Reply:
Correct. We actually add things like this throughout the codebase to help guide people that may want to call certain methods. Of course you'd just get a "undefined method" error normally, but this just lets us direct the user to use `destroy` instead.